doc/turnin.1
.TH TURNIN 1 "SEPTEMBER 2010" Turnin-NG
.SH NAME
turnin \- submit assignments to be graded
.SH SYNOPSYS
.B turnin [
.I OPTIONS
.B ] -c [
.I course-name
.B ] files
.SH DESCRIPTION
Submits assignments to be graded. We submit to the default project if it is not
specified as an option.

.SH REQUIRED ARGUMENTS
.IP "\-c, \-\-course"
Sets the course to which we'll submit our assignments.
.IP files
A list of files we wish to submit. They will be compressed using
.B tar(1)
and 
.B gzip(1)
formats before being copied to the submission directory.
.SH OPTIONS
.IP \-\-version
Print turnin's version
.IP "\-h, \-\-help"
Print a help message
.IP "\-l, \-\-list"
Prints a list of projects, along with whether or not they are enabled and shows
which project is the default project.
.IP "\-p, \-\-project"
Sets the project to which we'll submit our assignments.
.IP "\-C, \-\-config"
Path to an alternate configuration file
.IP "\-v, \-\-verbose"
Prints a list of submitted files once they have been submitted.
.IP "\-k, \-\-keyid"
Cryptographically signs an assignment using
.B gpg(1)
.
.IP "\-w, \-\-legal"
Print warranty and license information.

.SH EXAMPLES
.TP \w'turnin\ 'u
.BI turnin\ \-c \ world-cheeses \ \-\-list
List the projects in the course world-cheeses.
.TP
.BI turnin\ \-c \ world-cheeses\ stilton-recipe
Submit the file stilton-recipe to the default project in the course
world-cheeses.
.TP
.BI turnin\ \-c \ world-cheeses \ \-p \ stilton\ stilton-recipe
Submit the file stilton-recipe to the project stilton in the course
world-cheeses.

.SH FILES
.I /etc/turnin-ng.cf
.I ~/.turnin-ng/submissions

.SH SEE ALSO
.B turnincfg(1)

The full documentation for
.B turnin
is maintained as a Texinfo manual.  If the
.B info
and
.B turnin
programs are properly installed at your site, the command
.IP
.B info turnin-ng
.PP
should give you access to the complete manual.

.SH AUTHOR
Turnin-NG and this manpage were written by Ryan Kavanagh <rak@rak.ac>
in the summer of 2009 and are still actively maintained. Turnin-NG is a
replacement for the 'project' and 'turnin' commands written in 1990 for the
SPARC architecture by an unknown author.

Both Turnin-NG and this manpage are distributed under the terms of the GNU
General Public License version 2, or (at your option) any later version.


CHANGES
2.0 2020-12-19
 Global:
  * port to Python 3

1.3 2014-08-27
 turnin:
  * Create ~/.turnin-ng/submissions if it doesn't yet exist

1.2 2013-08-30
 Global:
 * Use errno module instead of raw values

 Documentation:
  * Fix broken texinfo
  * Switch HTML/PDF documentation building to use makeinfo

1.1 2010-09-01
 Global:
  * Ignore build files in Git (generated documentation, build/*, etc.)
  * Always initialize all of the course's options in the config files
  * Upgrade configuration files from old releases / installations. Preserves
    previous settings, just adds the options to the config files.
  * Accept that the legacy symlink may already exist from a previous install

 turnincfg:
  * Add support for group managed courses
  * Don't die if we receive an invalid username or group when creating courses
  * Don't strip a submission if it causes an overwrite
  * Added forgotten import (sys.exit() needs sys to be imported)
  * Don't crash when raising CalledProcessError when gpg --verify fails
  * Warn on dangling signatures
  * Warn on signatures submitted by other students
  * Set the group on the compressed tarball and extracted project directory
  * Also delete tarball on delete_project if project is compressed

 turnin:
  * Don't needlessly import chown
  * Fixed vulnerability. Set student's $HOME/.turnin-ng/list to 0600 so that, on
    off chance their home directory is world readable, other students can't find
    out what their submitted file suffixes are (and thus be able to play with
    them).
  * Print a pretty error message when we try to submit to a closed project
  * Don't crash when GPG fails to sign and we try to copy a non-existent .sig file
  * Remove stale signatures on unsigned resubmits

 Documentation:
  * Fix broken infodir entry
  * Properly install info manual with install-info
  * Expand the administrative section of the info manual

1.0.1 2010-02-12
 Global:
  * Fix typos in turnin and turnincfg option descriptions

 Documentation:
  * Fix typos in the Texinfo documentation and in manpages

1.0 2009-09-01
 Global:
  * The project command has been renamed to "turnincfg". We still install a link
    from $prefix/bin/project -> $prefix/bin/tcpsadmin , but we print a warning
    message when project is called and the link may disappear in any future
    release.  This was done following this discussion:
    http://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg677272.html
  * Rename turninng.sys to turninng.fileperms to avoid name clashes with
    Python's sys.

 turnin:
  * Elaborate the -v option in --help to avoid confusion

 project/tcpsadmin:
  * Project without any options calls --help
  * --list for projects works.
  * Archiving a course sets it 0600 and is owned by the course's user and group.
  * Added a workaround for python2.4 since its tar.tarfile doesn't have an
    extractall()
  * When disabling a course, check if it's default. If it is, unset the default
    course.
  * Print the course directory when asking if we want to delete it.

 Documentation:
  * Commented out '@allowcodebreaks false' since it breaks on older systems and
    the output remains the same.
  * Now build PDF/HTML/INFO pages and install them
  * Updated for the name change

1.0~rc1 2009-08-06
 Global:
  * Dropped missed references to check_group
  * Updated the docstrings

 project:
  * Substitute Error for Warning when stripping random suffixes.
  * [BUG FIX]: Check for a tarball, not a directory when we set the tarball key
    in the config
  * Check that ther's a course before creating a project in situations like
    'project -c foo apple', which creates the project apple in the course foo.
  * Simplify how we get the config file and the course from the user

 Documentation:
  * Update the INSTALL file to reference to the documentation for more
    information on configuring.
  * Add examples to the manpages

1.0~beta2 2009-08-04
 Global:
  * Get rid of 'No `START-INFO-DIR-ENTRY' and no `This file documents' when
    installing the info file.
  * If users don't pass any arguments, print --help message
  * Add --prefix note to the INSTALL file so sysadmins don't have to search
    online for how to install under /usr/local/
  * Install man pages in the right directories.
  * Project submission directory is 733 now, it would be too much of a hassle
    sysadmins to create a group per course and add students to a group per
    course. Drop group checking.
  * Split the configurations into a global config file (used by students and
    administrators) and a course config file (used by professors and TAs).
  * [BUG FIX]: When listing projects, check for a 'default' project. This avoids
    KeyErrors.
  * Rename config files to turnin-ng.cf
  * Add option to print warranty information, as suggested by GPL
  * Set the default configuration file in /etc/turnin-ng.cf
  * Rename the turnin package to turninng
 
 turnin:
  * Use random 16 character suffixes (username-XXXXXXXXXXXXXXXX.tar.gz) when
    submitting to prevent users from copying each other or overwriting each
    other's assignments.
  * Store these suffixes in $HOME/.turnin-ng/submissions and use UUIDs to
    identify each project. This is to prevent a course having the project A,
    removing it, and then recreating it. 
  * Have the user specify a project if the course doesn't have a default one.
 
 project:
  * Don't abort creating the course if the directory already exists.
  * [BUG FIX]: When run without args, we should not create a course and write to
    config file.
  * Initialize a project if we pass only a project name
  * Add turnin's list projects option.
  * When we compress projects, print the path to the tarball.
  * Add an option to compress a whole course directory.
  * Add an option to strip the random suffixes (-D / --perm-disable).
  * [BUG FIX]: Check that a project exists in a course before
    enabling/disabling/etc'ing it.

src/turnin.py
#!/usr/bin/python2.5
# Turnin-NG, an assignment submitter and manager. --- Turnin script
# Copyright (C) 2009  Ryan Kavanagh <ryanakca@kubuntu.org>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

from optparse import OptionParser, OptionGroup
import os.path
import sys

from turnin.configparser import TurninGlobal, TurninCourse, TurninProject
from turnin.submitter import submit_files, check_group, list_projects

if __name__ == '__main__':
    usage = '%prog [options] [project name]'
    parser = OptionParser(version='0.1_pre-alpha', usage=usage)
    parser.add_option('-c', '--course', help='Set the course to submit the ' +
            'assignment to.')
    parser.add_option('-l', '--list', help='Lists projects for the course. ' +
            'also displays wether or not the project is open.',
            action='store_true')
    parser.add_option('-p', '--project', help='Set the project to submit the '+
            'assigmnent to.')
    parser.add_option('-C', '--config', help='Set a custom configuration ' +
            'file.')
    parser.add_option('-v', '--verbose', action='store_true', help='Print ' +
            'a list of submitted files.')

    (options, args) = parser.parse_args()

    # If we're listing projects, it's obvious we won't be submitting to one.
    if not options.list and not args:
        sys.exit(ValueError("Error, please submit at least one document."))
    if not options.course:
        sys.exit(ValueError("Error, please specify a course."))
    try:
        # We're nesting this in a try: because we might find out the requested
        # course or project doesn't exist.
        if options.config:
            config = options.config
        else:
            config = os.path.join("/srv/submit", options.course, "turnin.cf")
        # We're listing the courses:
        if options.list:
            projects = list_projects(config, options.course)
            for i in projects:
                print i
            sys.exit()
    except ValueError, e:
        sys.exit(e)

    try:
        # We're creating after having run list because if the user is asking for
        # a list of projects, the user won't be passing a project.
        if options.project:
            project = TurninProject(config, options.course, options.project)
        else:
            project = TurninProject(config, options.course, TurninCourse(config,
                options.course).course['default'])
    except ValueError, e:
        sys.exit(e)


    # We have checked that the user has provided a course, a project and
    # assignments. We can now proceed to submit them.
    # Check that the current user is in the submitter group.
    if check_group(project.course['group']):
        files = submit_files(options.course, project, args)
        if options.verbose:
            print "Submitted files:"
            for file in files:
                print file
            sys.exit()
    else:
        sys.exit("You need to be in the group %s to submit files" %
                project.course['group'])
